src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Processors/GeneralizeProcess.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using EnsureThat;
+using Microsoft.Health.Fhir.Anonymizer.Core.Models;
+using Microsoft.Health.Fhir.Anonymizer.Core.Processors.Settings;
+using Microsoft.Health.Fhir.Anonymizer.Core.AnonymizerConfigurations;
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Model;
+using Hl7.FhirPath;
+using Newtonsoft.Json.Linq;
+using Hl7.Fhir.Model.Primitives;
+
+namespace Microsoft.Health.Fhir.Anonymizer.Core.Processors
+{
+
+    public partial class GeneralizeProcessor : IAnonymizerProcessor
+    {
+
+        private static readonly HashSet<string> s_integerValueTypeNames = new HashSet<string>
+        {
+            FHIRAllTypes.Integer.ToString(),
+            FHIRAllTypes.PositiveInt.ToString(),
+            FHIRAllTypes.UnsignedInt.ToString()
+        };
+
+        private static readonly HashSet<string> s_DateOrTimeValueTypeNames = new HashSet<string>
+        {
+            FHIRAllTypes.DateTime.ToString(),
+            FHIRAllTypes.Date.ToString(),
+            FHIRAllTypes.Time.ToString(),
+            FHIRAllTypes.Instant.ToString()
+        };
+
+        private void ValidateExpression(ElementNode node, string conditionExpression, string targetExpression)
```

Comment:
Looks like we do not need to validate twice.

Reply:
An expression could fail to parse, compile and predicate or scalar. My design is first validate parse and compile when loading the configuration file. Then, in each generalization process, validate predicate or scalar when we get the node information. 
If we only validate once, should we remove the validation of compile and directly validate predicate in each process?

Reply:
Parse, compile should be before real execution, we should through critical error to block customer.

Predict and scalar related to the runtime context, and it relates to the real data. I would prefer to consider it as a runtime error. Currently we provide options for customer to skip or stop for single record runtime error.

For this case, my major concern is that we actually call multiple predict and scalar at runtime. FHIR NetCore lib is not so efficient we should consider the performance. Here looks like anyway it is a runtime error, can we just throw runtime exception during processing? (instead of validate separately)

Reply:
Never mind, I saw your latest change.

Reply:
Yes, validate separately cause multiple predict and scalar. I fixed it that only predict and scalar one time.

Reply:
Thanks again for this good suggestion.